### PR TITLE
Kafka API changes for Swan lake

### PR DESCRIPTION
**Pull request by aashikam**
**Note: Some features are intentionally disabled since the APIs are not finalized for them**

Fixes: https://github.com/ballerina-platform/ballerina-standard-library/issues/672
Fixes: https://github.com/ballerina-platform/ballerina-standard-library/issues/637

## Purpose
Introducing the new API changes and service typing in the NATS Streaming module for Ballerina Swan Lake. The design docs are given below. 

1. [Kafka client API changes](https://docs.google.com/document/d/1O_RRExICxYxzBoOpmxayqBRuf_UGzF-4BxVuQ_M7MI8/edit?usp=sharing)
2. [Kafka service changes](https://docs.google.com/document/d/1AS1BD0qPdP1E870GKowwvcsoFb8ij9aS78X3y6Zop-Y/edit?usp=sharing) 

## Changes made
- No changes to kafka:Serializer/ kafka:Deserializer.
- kafka:Consumer is divided into kafka:Listener (asynchronous) and kafka:Consumer (synchronous). 

- [**Service-related changes**](https://docs.google.com/document/d/1AS1BD0qPdP1E870GKowwvcsoFb8ij9aS78X3y6Zop-Y/edit?usp=sharing)

## Examples with the added changes

**Producing a message**
```ballerina
import ballerinax/kafka;

kafka:ProducerConfiguration producerConfiguration = {
    bootstrapServers: "localhost:9092",
    clientId: "basic-producer",
    acks: "all",
    retryCount: 3,
    valueSerializerType: kafka:SER_BYTE_ARRAY,
    keySerializerType: kafka:SER_BYTE_ARRAY
};

kafka:Producer kafkaProducer = new (producerConfiguration);

string message = "Hello World, Ballerina";
kafka:ProducerError? result = kafkaProducer->send(message.toBytes(), "kafka-topic", key = 1);
```

**Subscribing and consuming messages**
```ballerina
kafka:ConsumerConfiguration consumerConfiguration = {
    bootstrapServers: "localhost:9092",
    groupId: "group-id",
    offsetReset: "earliest",
    topics: ["kafka-topic"]
};

listener kafka:Listener lis = new (consumerConfiguration);

service kafkaService on lis {
    // This resource will be executed when a message is published to the
    // subscribed topic/topics.
    remote function onMessage(kafka:Caler caller, kafka:ConsumerRecord[] records) {
    }
}
```

### CODE REVIEW

File: kafka-ballerina/listener.bal [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+    private isolated function subscribe(string[] topics) returns ConsumerError? {
+        if (self.consumerConfig?.groupId is string) {
+            return consumerSubscribe(self, topics);
+        } else {
+            panic createConsumerError("The groupId of the consumer must be set to subscribe to the topics");
+        }
+    }
+}
+
```
Comment (ThisaruGuruge) [inline kafka-ballerina/listener.bal:88]:
```suggestion
```

File: kafka-ballerina/types.bal [hunk 2]
```diff
@@ -24,3 +24,9 @@ public type AuthenticationMechanism AUTH_SASL_PLAIN;
 
 # Represents the supported security protocols for Kafka clients.
 public type SecurityProtocol PROTOCOL_PLAINTEXT|PROTOCOL_SASL_PLAINTEXT|PROTOCOL_SASL_SSL|PROTOCOL_SSL;
+
+# The Kafka service type
+public type KafkaService service object {
+    remote function onMessage(Caller caller, ConsumerRecord[] records);
+    // To be completed when support for optional params in remote functions is available in lang
+};
```
Comment (ThisaruGuruge) [inline kafka-ballerina/types.bal:32]:
```suggestion
};

```

File: kafka-ballerina/consumer.bal [hunk 3]
```diff
@@ -14,19 +14,14 @@
...
-import ballerina/lang.'object;
-
 # Represents a Kafka consumer endpoint.
 #
 # + consumerConfig - Used to store configurations related to a Kafka connection
 public client class Consumer {
-    *'object:Listener;
 
     public ConsumerConfiguration consumerConfig;
     private string keyDeserializerType;
```
[the remark this answers is not in the document]
Comment (aashikam) [reply kafka-ballerina/consumer.bal:23]: If we keep this, it is easier to add more serdes type support without breaking the code. 
I've only allowed BYTE_ARRAY for now. This has been accessed as a `field value` of `Listener` object, in all the places of java native code. Will fix them in a separate PR. 
https://github.com/ballerina-platform/module-ballerinax-kafka/blob/58bbd47009c33746f88cc8e84b3c1243834efd84/kafka-ballerina/types.bal#L6

**Review by ThisaruGuruge — APPROVED** [review]
LGTM

**Pull request merged**